README.md
# Response

Flutter UI package which will help you to build your responsive UI design as you wanted to be on any device as it recreates every widget to fit the device display size on which the app is working.


![thumbnail](https://github.com/AhmedAbouelkher/Response-for-Flutter/blob/master/screenshots/Combined.jpg)

### Features
- Easy to use, with so few but yet powerful methods.
- Fast, very reliable and readable.
- Saves you a lot of time trying to fit widget size to different displays.
- Works on any Android/ IOS device (including iPad).
- Works in Portrait mode or Landscape mode on the same device.


## How to use
**First** wrap your `MaterialApp` widget with `Response`.

```dart
     Response(
      originalScreenHeight: 759,
      originalScreenWidth: 392,
      child: MaterialApp(
        home: HomePage(),
      ),
    )
```
> **Note:** Responsive needs to be at the top of your **Widget Tree** to be able to work and rebulid the widgets to a new size as the device display changes.

> **Note:** if you want to use any functions features related to `MediaQuery` you **must** change the order to make the `Response` widget a child to the `MaterialApp` widget.

**Second**
declare an instance `response` of `ResponseUI()` class at the top of your file.

```dart
final response = ResponseUI.instance;
``` 

And that is all to be able to initialize the package, **Very easy!**


to use the package on your widget you have 4 easy to use methods:

| Name  | Description  |
| ------------ | ------------ |
| `response.setWidth(width)` |  Sets the widget width in pixels value to be relatively constant across all different display sizes. |
| `response.setHeight(height)` | Sets the widget height in pixels value to be relatively constant across all different display sizes. |
| `response.setFontSize(fontSize)` | Sets the text size in pixels to be relativly constant on all diplays sizes. |
| `response.isDevicePortrait` | Helpful if you want to know whether the device orientation is in Portrait or in Landscape.  |
| `response.inMobilePortrait` | Helps you if you want to know whether the device you are working on a Mobile Portrait (returns `true`) or a Tablet Portrait mode (returns `false`).  |
| `response.screenWidth` | returns the current device screen width as of type `double`. |
| `response.screenPixelRatio` | returns the current device screen ratio. (using `MediaQuery`)|
| `response.textScaleFactor` | returns the font scalling factore as of type `double`. (using `MediaQuery`)|
| `response.bottomPadding` | returns the current device screen padding offset from the bottom. (using `MediaQuery`)|
| `response.topPadding` | returns the current device screen padding offset from the top. (using `MediaQuery`)|

> ***Note:*** `isDevicePortrait` and `inMobilePortrait` returns a `bool` (true or false) with which you can know the current orientation state of
your device.

### Height and Width
to set a width and height to for example a Container.

```dart
//We want the Container Widget to be 200px in width and 200px in height
    Container(
        width: response.setWidth(300), //setting the Container width to be 300px
        height: response.setHeight(200), //setting the Container height to be 300px
        color: Colors.teal,
       )
```
**OR**

> ***Only*** for version 2.6.0 or above

```dart
//We want the Container Widget to be 200px in width and 200px in height
    Container(
        width: (300.0).asWidth(context), //setting the Container width to be 300px
        height: (200.0).asHieght(context), //setting the Container height to be 300px
        color: Colors.teal,
       )
```

> ***Note:*** the scaling process will work also in the same device in both
in Portrait mode or in Landscape mode.



### Text Size
to set the text size.

```dart
    Text(
        'This is a Test Text',
        style: TextStyle(
           fontSize: response.setFontSize(24), //defined a fixed font size in pixels
           fontWeight: FontWeight.bold,
    ),
)
```

**OR**

> ***Only*** for Flutter version 2.6.0 or above

```dart
    Text(
        'This is a Test Text',
        style: TextStyle(
           fontSize: (24.0).asFontSize(context), //defined a fixed font size in pixels
           fontWeight: FontWeight.bold,
    ),
)
```

### Getting Screen Width & Height

```dart
    TextSpan(
      text: '\nScreen Height: ${response.screenHeight}px\n', //we are getting back our current device screen 
      //height in pixels.
      style: TextStyle(
        color: Colors.black,
        fontSize: response.setFontSize(18),
      ),
    ),
    TextSpan(
      text: 'Screen Width: ${response.screenWidth}px', //we are getting back our current device screen 
      //width in pixels.
      style: TextStyle(
        color: Colors.black,
        fontSize: response.setFontSize(18),
      ),
    ),
```

## Contributions

If you feel like you can contribute to make the package much better and more useful for other
developers, don't hesitate to `fork` the repo and make `pull request` to your changes.

Source code on Github: [here](https://github.com/AhmedAbouelkher/Responsive-for-Flutter "here")


<!-- ## Some Tricks

If you feel confused or can't understand any part of the code, then the Package
Documentation will be very useful.

You can access any Method/Class Quick Documentation:
  in `VS Code`: By hovering with your cursor over the method/class in your code.

  in `Android Studio/IntelliJ IDEA`: move the cursor pointer to the method/class and type
  on your keyboard (`ctrl + q` on Windows/ `command + j` on MacOS).

You can access any Method/Class Full Detailed Documentation:
  in `VS Code` and `Android Studio/IntelliJ IDEA`:  hovering with your cursor over the method/class in your code while
  pressing (`ctrl` on Windows/ `command` on MacOS) button then click. -->

## Inspired
This package was inspired by [Prateek Sharma](https://github.com/PrateekSharma1712 "Prateek Sharma")'s great custom responsive UI [project](https://medium.com/flutter-community/flutter-responsive-ui-for-learning-platform-app-2df185f86e8e "project").


CHANGELOG.md
## [1.0.1] - 2021-4-6
* `Response` is now full `null-safe` 🎉🎉.
* Add new extensions `asHeight`, `asWidth` and `asFontSize` on `num` class.
* General Bug Fixes.


## [0.1.1] - 2020-5-30
* General Bug Fixes.
* General Performance improvments.

## [0.1.0] - 2020-5-17
* General Bug Fixes.

## [0.0.1+4] - 2020-5-9
* General Bug Fixes.
* General Performance Fixes.
* Fix scaling problem with tablets.
* Adding new MediaQuery related functions to decrease the developing time.
* Adding new extension features to increase performance.

## [0.0.1+3] - 2020-4-19
* Bug Fixes

## [0.0.1+2] - 2020-4-14
* Bug Fixes
* General Performance Fixes.

## [0.0.1+1] - 2020-4-8
* Bug Fixes

## [0.0.1] - 2020-4-1
* Simplified package initialization.
* Added a more readable/meaningful methods names.
* Adding a full documentation to all methods and classes.

lib/Response.dart
library response;

import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';

class Response extends StatelessWidget {
  ///[Response] is an initialization Class for the whole package.
  ///
  ///We are using it to get [constraints] and [orientation] information
  ///about the currently running app.

  Response(
      {@required this.child,
      this.originalScreenHeight = 759,
      this.originalScreenWidth = 392});

  ///`originalScreenHeight` is the original device screen height in pixels
  ///on which you designed your origial app layout.
  /// {@tool sample}
  ///
  /// ```dart
  /// Response(
  ///  originalScreenHeight: 759,
  ///  originalScreenWidth: 392,
  ///  child: MaterialApp(
  ///     home: HomePage(),
  ///  ),
  ///)
  /// ```
  /// {@end-tool}
  ///
  final double originalScreenHeight;

  ///`originalScreenWidth` is the original device screen width in pixels
  ///on which you designed your origial app layout.
  ///  /// {@tool sample}
  ///
  /// ```dart
  /// Response(
  ///  originalScreenHeight: 759,
  ///  originalScreenWidth: 392,
  ///  child: MaterialApp(
  ///     home: HomePage(),
  ///  ),
  ///)
  /// ```
  /// {@end-tool}
  ///
  final double originalScreenWidth;

  ///`child` is the child widget in which we want to use our package
  ///
  ///`child` must be in the top of our (Widget Tree) to work properly.
  ///You can say that your [child] widget is always will be our [MaterialApp]
  ///
  ///`child` is a requirment to enable the package to work.
  ///
  /// Still confused? Run the example app in which you can fine
  /// a working example with full comments to teach you every step
  /// the example: https://github.com/AhmedAbouelkher/UI_SizeConfig-for-Flutter/tree/master/example
  ///
  final Widget child;

  @override
  Widget build(BuildContext context) {
    ///[LayoutBuilder] is used to get the constraints/borders geometry of
    ///the device screen in which the app is working on.
    return LayoutBuilder(
      builder: (context, constraints) {
        ///[OrientationBuilder] is used to get the orientation mode of
        ///the device in which the app is working on.
        return OrientationBuilder(
          builder: (BuildContext context, Orientation orientation) {
            ///initializing our brain
            ResponseUI()._init(constraints, orientation, originalScreenHeight,
                originalScreenWidth);
            return child;
          },
        );
      },
    );
  }
}

///
///[ResponseUI] is the main brain of the package.
///
///We are using it to calculate the geometric value of our widgets
///in which we define for example a specific [width] or [height] in pixels
///and in that case we want the specified property to be constant whatever the
///device screen size would be.
///
///For example:
///if you want to specific a constant [height] and [width] to your container
///lets say it will be 100 pixels in height and 100 pixels in width, you will
///type the [height] property = 100 and the [width] property = 100, So in that
///case you want to maintain these [height] and [width] property relative to
///the device screen size you originally designed your app on.
///
/// in that case the class/package will do the work to change these relative
/// [height] and [width] to maintain the same display occupation size.
///
///

class ResponseUI {
  static double _screenWidth;
  static double _screenHeight;
  static double _blockWidth = 0;
  static double _blockHeight = 0;
  static bool _isPortrait = true;
  static bool _isMobilePortrait = false;
  static double _fixedHeightFactor = 0;
  static double _fixedWidthFactor = 0;

  void _init(BoxConstraints constraints, Orientation orientation,
      double originalHeight, double originalWidth) {
    if (orientation == Orientation.portrait) {
      _screenWidth = constraints.maxWidth;
      _screenHeight = constraints.maxHeight;
      _isPortrait = true;
      if (_screenWidth < 450) {
        _isMobilePortrait = true;
      }
    } else {
      _screenWidth = constraints.maxHeight;
      _screenHeight = constraints.maxWidth;
      _isPortrait = false;
      _isMobilePortrait = false;
    }
    _blockHeight = _screenHeight / 100;
    _blockWidth = _screenWidth / 100;
    _fixedHeightFactor = originalHeight / 100;
    _fixedWidthFactor = originalWidth / 100;
  }

  ///[isDevicePortrait] is helpful if you want to know whether the device orientation is
  ///in Portrait or in Landscape, it can ably to any device (Tablet or Mobil Phone).
  bool get isDevicePortrait => _isPortrait;

  ///[inMobilePortrait] is helpful if you want to know whether the device you are
  ///working on a Mobile Portrait or a Tablet Portrait mode and you can take action
  ///according to these information ot layout your UI.
  bool get inMobilePortrait => _isMobilePortrait;

  ///[screenWidth] returns the current device screen width.
  ///Note: You can use it insted of `MediaQuery.of(context).size.width`
  double get screenWidth => _screenWidth;

  ///[screenHeight] returns the current device screen height.
  ///Note: You can use it insted of `MediaQuery.of(context).size.height`
  double get screenHeight => _screenHeight;

  ///[setWidth] uses its argument [width] to calculate the initial widget width
  ///in pixels and from that the package can deal with all the calculation.
  ///`setWidth`
  /// {@tool sample}
  ///
  /// ```dart
  /// Container(
  ///   width: response.setWidth(300),
  ///   height: response.setHeight(200),
  ///   color: Colors.teal,
  ///  ), //Container
  /// ```
  /// {@end-tool}

  double setWidth(double width) {
    ///[width] is the widget width which you want it initially and finally be the same.
    if (_fixedWidthFactor < _blockWidth * 0.6) {
      _fixedWidthFactor *= 1.13;
    }
    return ((width / _fixedWidthFactor) * _blockWidth);
  }

  ///[setHeight] uses its argument [height] to calculate the initial widget height
  ///in pixels and from that the package can deal with all the calculation.
  ///`setHeight`
  /// {@tool sample}
  ///
  /// ```dart
  /// Container(
  ///   width: response.setWidth(300),
  ///   height: response.setHeight(200),
  ///   color: Colors.teal,
  ///  ), //Container
  /// ```
  /// {@end-tool}
  ///

  double setHeight(double height) {
    ///[height] is the widget height which you want it initially and finally be the same.
    return ((height / _fixedHeightFactor) * _blockHeight);
  }

  ///[setFontSize] uses its argument [fontSize] to calculate the initial text size
  ///in pixels and from that the package can deal with all the calculation.
  ///`setFontSize`
  /// {@tool sample}
  ///
  /// ```dart
  /// Text(
  ///    'This is a Test Text',
  ///     style: TextStyle(
  ///             fontSize: response.setFontSize(24),
  ///             fontWeight: FontWeight.bold,
  ///          ), //TextStyle
  ///       ) //Text
  /// ```
  /// {@end-tool}
  ///

  double setFontSize(double fontSize) {
    ///[fontSize] is the text size which you want it initially and finally be the same.
    return ((fontSize / _fixedHeightFactor) * _blockHeight);
  }
}
